### ISSUE TITLE
Some bugs of CategoricalCrossentropy and SparseCategoricalCrossentropy in tf.keras.losses 

### ISSUE THREAD

**Issue by gdhy9064** [tensorflow/tensorflow#38632]
It seems that there are some bugs of CategoricalCrossentropy and SparseCategoricalCrossentropy. The codes below will go wrong in tensorflow with version 2.2.0rc2.
1. Error of no gradients provided with categorical crossentropy .
```
import numpy as np
import tensorflow as tf
import tensorflow.keras as keras
import tensorflow.keras.backend as K
from tensorflow.keras.layers import *

i = Input([], dtype='int32')
o = Dense(2, activation='softmax')(i[:, None])
l = tf.keras.losses.categorical_crossentropy(K.one_hot(i, 2), o)
m = keras.Model(inputs=i, outputs=o)
m.add_loss(l)
m.compile(optimizer='adam')
m.fit(np.ones([1]))
```
and the same with the code below
```
import numpy as np
import tensorflow as tf
import tensorflow.keras as keras
import tensorflow.keras.backend as K
from tensorflow.keras.layers import *

i = Input([], dtype='int32')
o = Dense(2, activation='softmax')(i[:, None])
l = tf.keras.losses.CategoricalCrossentropy()(K.one_hot(i, 2), o)
m = keras.Model(inputs=i, outputs=o)
m.add_loss(l)
m.compile(optimizer='adam')
m.fit(np.ones([1]))
```
2. Showing error that 'Sparse ops are not supported with functional models with built-in layer wrapping...' with sparse categorical crossentropy.
```
import numpy as np
import tensorflow as tf
import tensorflow.keras as keras
import tensorflow.keras.backend as K
from tensorflow.keras.layers import *

i = Input([], dtype='int32')
o = Dense(2, activation='softmax')(i[:, None])
l = tf.keras.losses.sparse_categorical_crossentropy(i, o)
m = keras.Model(inputs=i, outputs=o)
m.add_loss(l)
m.compile(optimizer='adam')
m.fit(np.ones([1]))
```
as well as the one below
```
import numpy as np
import tensorflow as tf
import tensorflow.keras as keras
import tensorflow.keras.backend as K
from tensorflow.keras.layers import *

i = Input([], dtype='int32')
o = Dense(2, activation='softmax')(i[:, None])
l = tf.keras.losses.SparseCategoricalCrossentropy()(i, o)
m = keras.Model(inputs=i, outputs=o)
m.add_loss(l)
m.compile(optimizer='adam')
m.fit(np.ones([1]))
```

**Comment by amahendrakar** [comment]
The CategoricalCrossentropy sample code works with TF v2.1 but fails with TF v2.2.0-rc3 and TF-nightly. Whereas code for SparseCategoricalCrossentropy fails with all three versions. Please find the attached gist below
- TF v2.1
https://colab.research.google.com/gist/amahendrakar/9593cd3693f10da12f4c8eea9cecc735/38632-2-1.ipynb

- TF v2.2.0-rc3
https://colab.research.google.com/gist/amahendrakar/70de2b98f68054f90be8c1244f165946/38632-2-2.ipynb

- TF-nightly
https://colab.research.google.com/gist/amahendrakar/1399f73ff9092916d4c9ab408093e067/38632-tf-nightly.ipynb

Thanks!

**Comment by pavithrasv** [comment]
Thank you @gdhy9064. Could you provide more information on the use case you are looking to solve?

For the above use cases you can provide the loss in the compile API. If you want to create a loss that is dependent on layer inputs we recommend that you use `add_loss` API in layer `call`. Here is an example [colab](https://colab.sandbox.google.com/drive/1zzLcJ2A2qofIvv94YJ3axRknlA6cBSIw).

**Comment by gdhy9064** [comment]
Yes, I'd like to create a loss which is depandent on some layer inputs of a model. Sometimes I prefer to customize these losses with `add_loss` API outside of a model for convenience, though it maybe not be of a neat structure. In my case, I can simply use `tf.gather` instead of `sparse_categorical_crossentropy` to meet my needs, but it would be better if I can use `sparse_categorical_crossentropy` or `categorical_crossentropy` API directly. And many thanks for your guidances on colab.

**Comment by pavithrasv** [comment]
Thank you @gdhy9064, one simple workaround you can do for both the cases above is : remove activation parameter from the Dense layer and add a separate Activation layer like 

```
    o = layers.Dense(2)(i[:, None])
    o = layers.Activation('softmax')(o)
```

We will fix the underlying issue soon.

**Comment by gdhy9064** [comment]
Thank you @pavithrasv , it works well with `categorical_crossentropy`. It is a little pity that I can't still use with `sparse_categorical_crossentropy`.

**Comment by pavithrasv** [comment]
I made the change in this nightly [colab](https://colab.sandbox.google.com/gist/amahendrakar/1399f73ff9092916d4c9ab408093e067/38632-tf-nightly.ipynb#scrollTo=6vSINBWRA9GB). The workaround works for `sparse_categorical_crossentropy` as well. PTAL and let me know.

**Comment by gdhy9064** [comment]
Yes, it works for `sparse_categorical_crossentropy` in nightly version.
